File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    bool yield(vector<int>& bloomDay, int harvestDay, int m, int k){
        int adjacent = 0;
        for(auto b: bloomDay){
            adjacent = b<=harvestDay ? adjacent+1: 0;
            if(adjacent==k)
                adjacent=0, m--;    // made 1 bouquet
        }
        return m<=0;    // reqd no. of bouquets done
    }
    int minDays(vector<int>& bloomDay, int m, int k) {
        if((long)bloomDay.size()<(long)m*k)
            return -1;  // only defaulter case (since, if length>=m*k, it must be valid by infinity)
        auto lt=0, rt=1000000000;
        while(lt<rt){
            auto mid = lt+(rt-lt)/2;
            if(yield(bloomDay, mid, m, k))
                 rt=mid;
            else lt=mid+1;
        }
        return lt;
    }
};

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        int lt = 1, rt = Arrays.stream(bloomDay).max().orElseThrow(), ans = -1;
        while (lt <= rt) {
            int mid = lt + (rt - lt) / 2;
            if (solve(bloomDay, mid, m, k)) {
                ans = mid;
                rt = mid - 1;
            } else
                lt = mid + 1;
        }
        return ans;
    }

    private boolean solve(int[] bloomDay, int d, int m, int k) {
        int streak = 0, bouq = 0;
        for (int i : bloomDay) {
            if (i <= d) {
                if (++streak == k) {
                    streak = 0;
                    if (++bouq == m)
                        return true;
                }
            } else
                streak = 0;
        }
        return false;
    }
}